sale_pricelist_global_rule/models/sale_order.py

```diff
@@ -0,0 +1,73 @@
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    need_recompute_pricelist_global = fields.Boolean()
+    has_pricelist_global = fields.Boolean(compute="_compute_has_pricelist_global")
+
+    def _get_cummulative_quantity(self):
+        """Compute the cummulative quantity of products in the sale order.
+        :returns: dict{
+            by_template: {product.template: qty},
+            by_categ: {product.category: qty}}
+        }
+        """
+        self.ensure_one()
+        qty_data = {
+            "by_template": {},
+            "by_categ": {},
+        }
+        for line in self.order_line.filtered("product_id"):
+            qty_in_product_uom = line.product_uom_qty
+            # Final unit price is computed
+            # according to `qty` in the default `uom_id`.
+            if line.product_uom != line.product_id.uom_id:
+                qty_in_product_uom = line.product_uom._compute_quantity(
+                    qty_in_product_uom, line.product_id.uom_id
+                )
+            key_template = line.product_id.product_tmpl_id
+            key_categ = line.product_id.categ_id
+            qty_data["by_template"].setdefault(key_template, 0.0)
+            qty_data["by_template"][key_template] += qty_in_product_uom
+            qty_data["by_categ"].setdefault(key_categ, 0.0)
+            qty_data["by_categ"][key_categ] += qty_in_product_uom
+        return qty_data
+
+    @api.depends("pricelist_id")
+    def _compute_has_pricelist_global(self):
+        for sale in self:
+            if not sale.pricelist_id:
+                sale.has_pricelist_global = False
+                continue
+            qty_data = self._get_cummulative_quantity()
+            pricelist = sale.pricelist_id.with_context(
+                pricelist_global_cummulative_quantity=qty_data
+            )
+            suitable_rule = self.env["product.pricelist.item"]
+            for line in sale.order_line:
+                suitable_rule = pricelist._get_product_rule(
+                    line.product_id,
+                    quantity=line.product_uom_qty or 1.0,
+                    uom=line.product_uom,
+                    date=line.order_id.date_order,
+                )
+                if suitable_rule:
+                    break
+            sale.has_pricelist_global = bool(suitable_rule)
```

Comment:
I think the logic should be as follows, if I understand the intention correctly. Otherwise, `has_pricelist_global` becomes true for practically all cases.
```suggestion
            global_rule = self.env["product.pricelist.item"]
            for line in sale.order_line:
                suitable_rule = pricelist._get_product_rule(
                    line.product_id,
                    quantity=line.product_uom_qty or 1.0,
                    uom=line.product_uom,
                    date=line.order_id.date_order,
                )
                if not suitable_rule:
                    continue
                rule = global_rule.browse(suitable_rule)
                if rule.applied_on in ["3_1_global_product_template", "3_2_global_product_category"]:
                    global_rule = rule
                    break
            sale.has_pricelist_global = bool(global_rule)
```

Reply:
I don't understand your change, the current logic looks fine to me.
The intention is to activate the field when any sale line has a product with a global pricelist that can be applied.
Note that the `_is_applicable_for` function handles this logic:
 https://github.com/factorlibre/sale-workflow/blob/5d1e29774a3986653a456c72e2c69581ec6246be/sale_pricelist_global_rule/models/product_pricelist.py#L158C42-L163